mobile/apps/photos/lib/services/smart_albums_service.dart

```diff
@@ -0,0 +1,180 @@
+import "package:photos/ui/actions/collection/collection_sharing_actions.dart";
+
+class SmartAlbumsService {
+  final _logger = Logger((SmartAlbumsService).toString());
+
+  int _lastCacheRefreshTime = 0;
+
+  Future<Map<int, SmartAlbumConfig>>? _cachedConfigsFuture;
+
+  void clearCache() {
+    _cachedConfigsFuture = null;
+    _lastCacheRefreshTime = 0;
+  }
+
+  int lastRemoteSyncTime() {
+    return entityService.lastSyncTime(EntityType.smartAlbum);
+  }
+
+  Future<Map<int, SmartAlbumConfig>> getSmartConfigs() async {
+    final lastRemoteSyncTimeValue = lastRemoteSyncTime();
+    if (_lastCacheRefreshTime != lastRemoteSyncTimeValue) {
+      _lastCacheRefreshTime = lastRemoteSyncTimeValue;
+      _cachedConfigsFuture = null; // Invalidate cache
+    }
+    _cachedConfigsFuture ??= _fetchAndCacheSaConfigs();
+    return await _cachedConfigsFuture!;
+  }
+
+  Future<Map<int, SmartAlbumConfig>> _fetchAndCacheSaConfigs() async {
+    _logger.finest("reading all smart configs from local db");
+
+    final entities = await entityService.getEntities(EntityType.smartAlbum);
+
+    final result = _decodeSaConfigEntities({"entity": entities});
+
+    return result;
+  }
+
+  Map<int, SmartAlbumConfig> _decodeSaConfigEntities(
+    Map<String, dynamic> param,
+  ) {
+    final entities = (param["entity"] as List<LocalEntityData>);
+
+    final Map<int, SmartAlbumConfig> saConfigs = {};
+
+    for (final entity in entities) {
+      try {
+        final config = SmartAlbumConfig.fromJson(
+          json.decode(entity.data),
+          entity.id,
+          entity.updatedAt,
+        );
+
+        saConfigs[config.collectionId] = config;
+      } catch (error, stackTrace) {
+        _logger.severe(
+          "Failed to decode smart album config",
+          error,
+          stackTrace,
+        );
+      }
+    }
+
+    return saConfigs;
+  }
+
+  Future<void> syncSmartAlbums() async {
+    final cachedConfigs = await getSmartConfigs();
+
+    for (final entry in cachedConfigs.entries) {
+      final collectionId = entry.key;
+      final config = entry.value;
+
+      final infoMap = config.infoMap;
+
+      // Person Id key mapped to updatedAt value
+      final updatedAtMap = await entityService.getUpdatedAts(
+        EntityType.cgroup,
+        config.personIDs.toList(),
+      );
+
+      for (final personId in config.personIDs) {
+        // compares current updateAt with last added file's updatedAt
+        if (updatedAtMap[personId] == null ||
+            infoMap[personId] == null ||
+            (updatedAtMap[personId]! <= infoMap[personId]!.updatedAt)) {
+          continue;
+        }
+
+        final toBeSynced = (await SearchService.instance
+                .getClusterFilesForPersonID(personId))
+            .entries
+            .expand((e) => e.value)
+            .toList()
+          ..removeWhere(
+            (e) =>
+                e.uploadedFileID == null ||
```

Comment:
Will this only add files that are owned by the current user and are not hidden?

Reply:
Yes